python_modules/dagster/dagster/_config/pythonic_config/__init__.py

```diff
@@ -904,16 +924,56 @@ def with_resource_context(
         return copy
 
     def _initialize_and_run(self, context: InitResourceContext) -> TResValue:
-        updated_resource = (
-            self._resolve_and_update_nested_resources(context)  # noqa: SLF001
-            .with_resource_context(context)
-            ._resolve_and_update_env_vars()
-        )
+        with self._resolve_and_update_nested_resources(context) as has_nested_resource:
+            updated_resource = has_nested_resource.with_resource_context(  # noqa: SLF001
+                context
+            )._resolve_and_update_env_vars()
+
+            updated_resource.setup_for_execution(context)
+            return updated_resource.create_resource(context)
 
-        return updated_resource._create_object_fn(context)  # noqa: SLF001
+    @contextlib.contextmanager
+    def _initialize_and_run_cm(
+        self, context: InitResourceContext
+    ) -> Generator[TResValue, None, None]:
+        with self._resolve_and_update_nested_resources(context) as has_nested_resource:
+            updated_resource = has_nested_resource.with_resource_context(  # noqa: SLF001
+                context
+            )._resolve_and_update_env_vars()
+
+            with updated_resource.yield_for_execution(context) as value:
+                yield value
+
+    def setup_for_execution(self, context: InitResourceContext) -> None:
+        """Optionally override this method to perform any pre-execution steps
+        needed before the resource is used in execution.
+        """
+        pass
+
+    def teardown_after_execution(self, context: InitResourceContext) -> None:
+        """Optionally override this method to perform any post-execution steps
+        needed after the resource is used in execution.
+
+        teardown_after_execution will be called even if the resource fails to initialize in the
+        setup_for_execution step, and if any part of the run fails.
+        """
+        pass
 
-    def _create_object_fn(self, context: InitResourceContext) -> TResValue:
-        return self.create_resource(context)
+    @contextlib.contextmanager
+    def yield_for_execution(self, context: InitResourceContext) -> Generator[TResValue, None, None]:
+        """Optionally override this method to perform any lifecycle steps
+        before or after the resource is used in execution. By default, calls
+        setup_for_execution before yielding, and teardown_after_execution after yielding.
+
+        Note that if you override this method and want setup_for_execution or
+        teardown_after_execution to be called, you must invoke them yourself.
+        """
+        try:
+            self.setup_for_execution(context)
+            yield self.create_resource(context)
+        finally:
+            # what to do if teardown_after_execution throws
+            self.teardown_after_execution(context)
```

Comment:
So I'm going to go ahead and argue the case that `setup_for_execution` should be invoked outside of the `try`. That would mean that `teardown_after_execution` would _not_ be called if `setup_for_execution` throws. 

In effect this would say that if `setup_for_execution` fails, the resource is in an invalid, untrusted state and will not be invoked by the framework.

On a practical basis, this prevents one from having to write a bunch of boilerplate in teardown methods to prevent crashes if the setup fails.

Imagine the following class:
```
class ResourceWithComplexSetup(ConfigurableResource):
    # ...
    def setup_for_execution(self):
        self._thing_one = get_thing_one()
        self._thing_two = get_thing_two()
```

Now imagine that and `get_thing_one`, and `get_thing_two` can both throw.

Now the user has to deal with many states of the object. Namely that both `_thing_one` and `_thing_two` are None (`get_thing_one` threw), `_thing_two` is None ('get_thing_two' threw), or neither is None

In order to do thing safely at minimum you have to ensure that each object is not None to account for the three possible states of the object.
```
class ResourceWithComplexSetup(ConfigurableResource):
    # ...
    def teardown_after_execution(self):
        if self._thing_one:
            self._thing_one.cleanup()
        if self._thing_two:
           self._thing_two.cleanup()
```
This obviously can be much more complex. There are more subtle cases in this when the code in the setup methods are modifying some global state in some way. You have to think about the combination of all the possible codepaths and think about all the possible states the object might be in to handle teardown safely 

It's worth noting that we have an escape hatch in _either_ case, in that we can instruct the user to override `yield_for_execution` if the default we choose does not suit them.  They can put setup code inside or outside of the `try` block they write themselves. (The purist in me wants to only have this single plug-ability point, to simplify the relationship between framework code and user code [avoiding the conversation about the semantics of teardown entirely], but it would make the simple cases a bit annoying).

Reply:
This seems reasonable to me. Updated.